diff -up smartmontools-7.1/scsiprint.cpp.json-pending-count smartmontools-7.1/scsiprint.cpp
--- smartmontools-7.1/scsiprint.cpp.json-pending-count	2026-08-10 11:23:11.474375067 +0200
+++ smartmontools-7.1/scsiprint.cpp	2026-08-10 11:25:21.811866906 +0200
@@ -524,7 +524,7 @@ scsiPrintPendingDefectsLPage(scsi_device
         pl = bp[3] + 4;
         switch (pc) {
         case 0x0:
-            printf("  Pending defect count:");
+            jout("  Pending defect count:");
             if ((pl < 8) || (num < 8)) {
                 print_on();
                 pout("%s truncated descriptor\n", pDefStr);
