Index: trunk/smartmontools/scsiprint.cpp
===================================================================
--- smartmontools/scsiprint.cpp	(revision 5076)
+++ smartmontools/scsiprint.cpp	(revision 5090)
@@ -2340,6 +2340,6 @@
                      !wce ? "Disabled" : "Enabled");
         }
-    } else
         any_output = true;
+    }
 
     if (options.drive_info)
@@ -2463,12 +2463,10 @@
         if (gTempLPage)
             scsiPrintTemp(device);
-    }
-    // in the 'smartctl -a" case only want: "Accumulated power on time"
-    if ((! options.smart_background_log) && is_disk) {
-        if (! checkedSupportedLogPages)
-            scsiGetSupportedLogPages(device);
-        res = 0;
-        if (gBackgroundResultsLPage)
-            res = scsiPrintBackgroundResults(device, true);
+        // in the 'smartctl -A' case only want: "Accumulated power on time"
+        if ((! options.smart_background_log) && is_disk) {
+            res = 0;
+            if (gBackgroundResultsLPage)
+                res = scsiPrintBackgroundResults(device, true);
+        }
         any_output = true;
     }
